cpp/src/arrow/acero/scalar_aggregate_node.cc

```diff
@@ -240,8 +240,8 @@ Status ScalarAggregateNode::InputReceived(ExecNode* input, ExecBatch batch) {
     // We add segment to the current segment group aggregation
     auto exec_batch = full_batch.Slice(segment.offset, segment.length);
     RETURN_NOT_OK(DoConsume(ExecSpan(exec_batch), thread_index));
-    RETURN_NOT_OK(
-        ExtractSegmenterValues(&segmenter_values_, exec_batch, segment_field_ids_));
+    RETURN_NOT_OK(ExtractSegmenterValues(&GetLocalState()->segmenter_values, exec_batch,
+                                         segment_field_ids_));
```

Comment:
`OutputResult` seems non-thread safe, how can `InputReceived` be called from several threads at once? Am I missing something?

Reply:
> `OutputResult` seems non-thread safe, 

True.

> how can `InputReceived` be called from several threads at once?

`InputReceived` can't be parallel if there exists any segment keys [1]. So if `InputReceived` is parallel, then the segmenter will be `NoKeysSegmenter` which only generates "open and extending" segments, which in turn never triggers `OutputResult`.

[1] https://github.com/apache/arrow/blob/fc0862a25fa2d6aa17df14daa9ad37430d432425/cpp/src/arrow/acero/scalar_aggregate_node.cc#L178-L180

Reply:
Ah, so the data race occurred in the non-segmented case? It's weird that we have to change the segmenting state to thread-local to fix that :)

Reply:
It is weird and I may even withdraw this fix. Please wait for my answer for the other comment (it's long and I'm still writing). Thank you.